### PR TITLE
[MAINT] fix failure windows with Python 3.9 and matplotlib==3.3.0

### LINKED ISSUE #4720: failures on windows-latest with Python 3.9 and matplotlib==3.3.0
We are getting some reliable CI failures on main with the workflow that test our oldest dependencies.

It affects windows with python 3.9 and matplotlib==3.3.0 (all the other dependencies should be the most recent version).

See last week:
https://github.com/nilearn/nilearn/actions/workflows/testing_minimum.yml?query=is%3Afailure+branch%3Amain

As far I can tell the test session contains several warning like:

```
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts\pytest.exe\__main__.py", line 7, in <module>
    sys.exit(console_main())
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 201, in console_main
    code = main()
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 156, in main
    config = _prepareconfig(args, plugins)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 341, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 1140, in pytest_cmdline_parse
    self.parse(args)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 1494, in parse
    self._preparse(args, addopts=addopts)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 1398, in _preparse
    self.hook.pytest_load_initial_conftests(
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 1222, in pytest_load_initial_conftests
    self.pluginmanager._set_initial_conftests(
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 581, in _set_initial_conftests
    self._try_load_conftest(
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 619, in _try_load_conftest
    self._loadconftestmodules(
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 659, in _loadconftestmodules
    mod = self._importconftest(
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 710, in _importconftest
    mod = import_path(
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\pathlib.py", line 582, in import_path
    importlib.import_module(module_name)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\assertion\rewrite.py", line 184, in exec_module
    exec(co, module.__dict__)
  File "D:\a\nilearn\nilearn\nilearn\conftest.py", line 27, in <module>
    if is_matplotlib_installed():
  File "D:\a\nilearn\nilearn\nilearn\_utils\helpers.py", line 268, in is_matplotlib_installed
    import matplotlib  # noqa: F401
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\matplotlib\__init__.py", line 174, in <module>
    _check_versions()
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\matplotlib\__init__.py", line 159, in _check_versions
    from . import ft2font
AttributeError: _ARRAY_API not found
```

Also leads to 2 tests failures:

```
FAILED nilearn/tests/test_package_import.py::test_import_reporting_should_raise_warning_if_matplotlib_not_installed - ImportError: numpy.core.multiarray failed to import
FAILED nilearn/tests/test_package_import.py::test_import_plotting_should_raise_warning_if_matplotlib_not_installed - ImportError: numpy.core.multiarray failed to import
```

### PR DESCRIPTION
**Pull request by Remi-Gau**
- closes #4720
- try to find minimal version of matplotlib that fixes this
  - matplotlib 3.4: same warning / failure
  - matplotlib 3.5: same warning / failure
  - matplotlib 3.6: same warning / failure
  - matplotlib 3.7: same warning / failure
  - matplotlib 3.8: fine

### CODE REVIEW

**Comment by Remi-Gau** [comment]
OK I think that for this one we won't have another choice but to let the few users on windows handle this one themselves either by downgrading numpy to < 2.0 or upgrading matplotlib to the most recent version.

**Comment by bthirion** [comment]
It's a temporeary incompatibility btw versions ?

**Comment by Remi-Gau** [comment]
> It's a temporeary incompatibility btw versions ?

pretty much but it seems to only affect users who would:
- use windows
- use numpy >= 2.0
- use an "old version" of matplotlib

**Comment by bthirion** [comment]
We're probably not the only ones to face that.  But I understand it's a cornercase.

**Comment by Remi-Gau** [comment]
not sure how much we can make of this, but seems most of our downloads are on linux and a decent chunk of downloads are still on python 3.9.

**Comment by bthirion** [comment]
I guess we skip the failing test when matplotlib < 3.8 AND numpy >= 2

**Comment by Remi-Gau** [comment]
ok will do that then

File: .github/workflows/testing_minimum.yml
Comment (Remi-Gau) [inline .github/workflows/testing_minimum.yml:16]:
```suggestion
```

**Review by bthirion — APPROVED** [review]
Need to wait for CI to complete but LGTM

**Comment by Remi-Gau** [comment]
This is fairly small so will merge with one review only.

**Comment by hndgzkn** [comment]
> I guess we skip the failing test when matplotlib < 3.8 AND numpy >= 2

Sorry, I've seen this a bit late @Remi-Gau and @bthirion . 

I would suggest an alternative fix for this in another PR.

-  we might [set the min matplotlib version](https://peps.python.org/pep-0631/#example) for windows platform, or

- we can catch this in the code and give a proper warning to user. 
 
Otherwise tests will pass in CI, however a user who has such an installation will have to find the solution himself. 

As far as I understand, a part from matplotlib - numpy version incompatibility, there is also another problem in matplotlib version for windows, see https://github.com/matplotlib/matplotlib/issues/18292/

**Comment by Remi-Gau** [comment]
> * we might [set the min matplotlib version](https://peps.python.org/pep-0631/#example) for windows platform,

but it's not just a windows issue: it's windows **+ numpy >= 2**, not sure we can have this kind of condtional in pyproject.toml

**Comment by Remi-Gau** [comment]
also the user will get a warning like this one send by numpy

```
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.
```

So I think it is enough to put most people on track to find the answer (if they read errors).

**Comment by hndgzkn** [comment]
> > * we might [set the min matplotlib version](https://peps.python.org/pep-0631/#example) for windows platform,
> 
> but it's not just a windows issue: it's windows **+ numpy >= 2**, not sure we can have this kind of condtional in pyproject.toml

I just could not be sure, it might at first place be an issue with matplotlib version and windows. 
See https://github.com/matplotlib/matplotlib/issues/18292

And actually, we still get the error many times, but the tests do not fail. Check [this one](https://github.com/nilearn/nilearn/actions/runs/11846566381/job/33014479855).

May be the tests would fail if there was a test with matplotlib version 3.3 and numpy < 2, but there is no such test.

**Comment by Remi-Gau** [comment]
- Started a run in https://github.com/nilearn/nilearn/pull/4729 to check it

**Comment by Remi-Gau** [comment]
no failures and no warning that I could see in this run

https://github.com/nilearn/nilearn/actions/runs/11859585510/job/33052869109?pr=4729#step:4:57

**Comment by hndgzkn** [comment]
> no failures and no warning that I could see in this run
> 
> https://github.com/nilearn/nilearn/actions/runs/11859585510/job/33052869109?pr=4729#step:4:57

Ok, thank you for trying this, and sorry for taking your time. If you think the displayed message is ok, then ok with me.

**Comment by Remi-Gau** [comment]
That was fairly quick to check (especially now that our test runs are much faster) and better make sure that we had also done that check.

I would say that we have done "enough" on our end. Some users may face some issues but I would expect it to be small.
So I am just being reactive on this one and only do something if we get a deluge of complaints and issues.

Though in general the answer will be
- upgrade matplotlib 
- or downgrade numpy

**Pull request merged**
